pandas/_libs/tslibs/offsets.pyx

```diff
@@ -418,6 +509,31 @@ cdef inline int month_add_months(pandas_datetimestruct dts, int months) nogil:
     return 12 if new_month == 0 else new_month
 
 
+def shift_day(datetime other, int days):
+    """
+    Increment the datetime `other` by the given number of days, retaining
+    the time-portion of the datetime.  For tz-naive datetimes this is
+    equivalent to adding a timedelta.  For tz-aware datetimes it is similar to
+    dateutil's relativedelta.__add__, but handles pytz tzinfo objects.
+
+    Parameters
+    ----------
+    other : datetime or Timestamp
+    days : int
+
+    Returns
+    -------
+    shifted: datetime or Timestamp
+    """
+    if other.tzinfo is None:
+        return other + timedelta(days=days)
+
+    tz = other.tzinfo
+    naive = other.replace(tzinfo=None)
+    shifted = naive + timedelta(days=days)
+    return _localize_pydatetime(shifted, tz)
```

Comment:
so I think we ought to de-privatize _localize_pydatetime (do this now) then and in future cimport this